WordPress/src/main/java/org/wordpress/android/ui/reader/repository/usecases/PostLikeActionUseCase.kt

```diff
@@ -0,0 +1,84 @@
+package org.wordpress.android.ui.reader.repository.usecases
+
+import kotlinx.coroutines.CoroutineDispatcher
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode.BACKGROUND
+import org.wordpress.android.models.ReaderPost
+import org.wordpress.android.modules.IO_THREAD
+import org.wordpress.android.ui.reader.actions.ReaderPostActions
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryEvent
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryEvent.PostLikeEnded
+import org.wordpress.android.util.EventBusWrapper
+import javax.inject.Inject
+import javax.inject.Named
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class PostLikeActionUseCase @Inject constructor(
+    @Named(IO_THREAD) private val ioDispatcher: CoroutineDispatcher,
+    private val eventBusWrapper: EventBusWrapper
+) : ReaderRepositoryDispatchingUseCase(ioDispatcher) {
+    private val continuations: MutableMap<PostLikeRequest, Continuation<ReaderRepositoryEvent>?> = mutableMapOf()
+
+    init {
+        eventBusWrapper.register(this)
+    }
+
+    suspend fun perform(
+        post: ReaderPost,
+        isAskingToLike: Boolean,
+        wpComUserId: Long
+    ): ReaderRepositoryEvent {
+        val request = PostLikeRequest(post.postId, post.blogId, isAskingToLike, wpComUserId)
+
+        if (continuations[request] != null) {
+            throw IllegalStateException("Perform Like action has already been sent for this request")
```

Comment:
I might have been the one who proposed throwing an exception here. However, I'm a bit worried it might bite us. Imagine the user keeps clicking on "like->unlike -> like -> unlike". The app might crash with this exception, right?

Reply:
> Imagine the user keeps clicking on "like->unlike -> like -> unlike". 

@malinajirka I tested that scenario and it works as I expected. It stops a duplicate request from happening and did not crash. Now if that is not the required behavior, we can turn the map into a queue and process each request in order, so that we can match each request with the proper continuation.

Reply:
Ohh I understand it now. We are catching the exception in the layer above. In that case the issue is somewhere else. Is there any reason why we are throwing an exception for expected(valid) user flow? Moreover, the current design forces you to know the implementation of PostLikeActionUseCase when you want to use it which might not be ideal, wdyt?
I'd consider logging a debug message here and returning from the method instead of throwing an exception, wdyt?

Reply:
@malinajirka Yeah, I just poked around in stats and they seem do similar, but I am going to get rid of the exception, because pressing the button a bazillion times doesn't mean it's wrong. I wonder if what we need is a small delay before kicking off the request to the repo  - since the UI is optimistic? If we are sending these requests async, can we really guarantee the order of execution and even if that, we can't guarantee the order of return on the bus either. I'll need to change a timestamp to the request object to make it unique (mapping request/response)   wdyt?

Reply:
>can we really guarantee the order of execution

Nope, we can't easily guarantee that. Tbh I wouldn't worry about it. 

What we could theoretically do is show a snackbar when the user initiates opposite (like vs unlike) action while the previous action is still ongoing and ignore the user request. But I'm not sure it's worth it. Wdyt?

Reply:
>But I'm not sure it's worth it.

I don't think it's worth it. Let's go with the original intent, but remove the exception. If you send a duplicate request when one is running, it gets tossed. thumbs up or thumbs down?